### PR TITLE
Support editable installs of Cython pxd files

### LINKED ISSUE #442: Support editable installs of packages containing Cython declaration files
## Problem statement
I don't understand enough about how scikit-build-core is working here to be confident in any recommendation, so to avoid the XY problem I'll state exactly what I'm observing before I make any suggestions.

I have two projects that use Cython where one contains pxd files with declarations that I want to be able to `cimport` into the other package. If I install the first package, everything works fine when I build the second package. If I have an editable installation of the first package, however, Cython fails to find the headers:

```
  Error compiling Cython file:
  ------------------------------------------------------------
  ...
  from foo.bar_a cimport baz
  ^
  ------------------------------------------------------------

  ...: 'foo/bar_a.pxd' not found
```

I would like for this to work.

## Some observations
IIUC the way editable installs work is that build artifacts are copied into site-packages from the build directory (ideally not a temp one, so something specified like `build-dir=build/{wheel_tag}` in pyproject.toml) after the build completes, but Python modules are instead remapped via a custom import loader registered via `*_editable.py` in site-packages.py. 

Inspecting this list reveals that Cython modules are being included. I see something like this in the installed `_editable.py` file:
```
install({'foo.bar_a': '/path/to/foo/bar_a.pyx', 'foo.bar_b': '/path/to/foo/bar_b.pxd'}, {'foo.bar_a': 'foo/bar_a/lib.cpython-310-x86_64-linux-gnu.so'}, '$build_dir', False, True, [], [])
```
In this case, I've introduce an additional module `foo.bar_b` that is only a set of (in this case `cdef extern`) declarations. That could be useful for the purpose of e.g. doing what Cython's C/C++ standard library support does with `cimport`ing from `libc` or `libcpp`.

Note that, interestingly, the pyx file is included in the set of what I assume are source files, but the corresponding compiled extension module is also included in the second list of what I assume are meant to be importable extension modules (and the extension module is not in the source tree, it is placed into site-packages). On the other hand, the pxd file shows up in the first list and not in the second.

My guess is that the pyx file shouldn't be in the first list at all; Cython source files require compilation, so there's no reason to include them in the first list if it is indeed a list of source files that should be found in the search tree upon import. However, I don't think that's relevant for my problem, and I suspect that in practice it has no effect and importlib just ignores the pyx file. What I'm observing is because Cython's cimports don't have anything to do with Python's import machinery, so registering this import loader for a pxd file has no effect on Cython compilation.

## Proposed solution
If my ramblings above seem correct, then my guess is that the best solution here would be for editable installs to support some way to specify an additional set of files that need to be installed into the site directory for an editable install. It is insufficient to only take extension modules when Cython headers are in use, so I think the pxd files need to be installed as well just like the DSOs. My suggestion of configuring an additional set of files (ideally globs or so) is to avoid indexing too heavily towards Cython and to instead make the support more general.

### PR DESCRIPTION
**Pull request by vyasr**
Editable installs of packages containing Cython currently do not properly support downstream Cython packages finding the pxd files from the editable install. The issue is that editable installs are managed using a meta path finder, and [Cython does not appear to support that for pxd files](https://github.com/cython/cython/issues/5855). However, Cython does support pth files adding include paths directly, so this PR solves the pxd search issue by adding the necessary paths directly to the pth file that currently adds the meta path finder.

Resolves #442

### CODE REVIEW

**Comment by vyasr** [comment]
@LecrisUT @henryiii here's my first pass at working on #442. I'm running into an issue with it that I can't quite understand though. The scaffolding for the MRE is in place, but for some reason I get a build failure of the first package (pkg1) when I turn off build isolation (see https://github.com/scikit-build/scikit-build-core/pull/516/files#diff-d5c95d2fd1275d136b05a4336ce0df3dea361e1003672881d6068d5e05a6385dR56). @henryiii would you have an idea what that's about? I suspect that it's a paths issue because I see output indicating that the Cython compilation succeeded, but then the actual C compilation step fails saying that the `one.c` file is missing, which to me indicates that something is wrong with the paths and it's being put in the wrong place. When I inspect the temp directories pytest produced for the test I don't find the Cython-generated C file anywhere. However, the isolated build of pkg1 works just fine.

Once we sort that out, uncommenting the installation of pkg2 below that should demonstrate the issue. I've parametrized the test on editability, so in theory we should see one passed test (when editable mode is off) and one failed test (when editable mode is on).

The predefined pytest fixtures aren't really designed for multiple packages being installed in a single test, so I had to import some stuff manually from conftest to produce a reasonable test. Happy to look into cleaner solutions later, but for now I don't think it's critical for an initial demo of the problem.

To make it easier to see the differences between my PR and @LecrisUT's original that this is based off of (#399) I've opened a corresponding PR from my branch onto his at https://github.com/LecrisUT/scikit-build-core/pull/1. That PR can be used to view just the relevant diff, while this PR will run CI for us.

**Comment by vyasr** [comment]
OK weird, I found the issue although I don't quite understand the underlying cause. It seems like there's a bug where it works with `ninja` but not `make`? Installing `ninja` into the environment allows the unisolated build to run. On the surface that sounds like a CMake bug, but given that it has to do with Python editable installs and build isolation maybe it's a situation where scikit-build-core needs to pass some extra information to CMake when the generator is Unix Makefiles.

**Comment by vyasr** [comment]
@LecrisUT https://github.com/scikit-build/scikit-build-core/pull/516/commits/64b615d3e07ca594a1fe593c2d3668cc5823a0b0 shows the error arising with the original layout, while https://github.com/scikit-build/scikit-build-core/pull/516/commits/31c27ccc9e8c3556c69af712b86e9aea9ab6227d shows that the same error appears when switching to a `src/` layout for both packages. Locally I've tested rebasing out your changes from #399 and I see the same error as well.

**Comment by LecrisUT** [comment]
Quick note about `src-layout`, it's only relevant tothe original source files. The installed files hqve to strip out the `src/` prefix. Built files don't care about src format either.

**Comment by henryiii** [comment]
#399 is in, but it doesn't work if there's an `__init__.py` or on Python 3.9. So this probably won't be working yet.

**Comment by vyasr** [comment]
A bit swamped this week but will try to get back into this in the next week or two.

**Comment by vyasr** [comment]
@henryiii OK, when you get a chance could you give me some pointers on how you think this feature should be implemented? You seemed to have some thoughts on #442 about what we should and shouldn't do that I didn't follow.

**Comment by vyasr** [comment]
Depending on your level of familiarity with Cython, we may need to get some help from a Cython developer here. From a quick inspection, it seems like since Cython is essentially a preprocessing transpiler of the pyx files, I'm not sure whether `cimport` statements respect meta path modifications done via the mechanisms used by the `_$pkg_editable.py` file. Naively I might have expected them to do so, but if I add some print statements inside `find_spec` I'm not seeing them during the Cython compilation step.

Furthermore, pxd files would change the 1-1 mapping from module to source file into a many-to-one mapping. I've done that by manually modifying the installed `_pkg1_editable.py` file and it works in the specific test case, but it makes the behavior ill-defined if `find_spec` goes into the `if fullname in self.known_source_files` block and then hgas to only return a single spec from a list of files, which doesn't make sense. We need some way to differentiate pxd from py files, and since `find_spec` doesn't provide that it makes me extra skeptical that this is the right mechanism for making pxd files searchable.

On the issue you also commented

> I'm pretty sure you don't want to "install" the header files, you just want to be able to find them in-place. Similarly, I'd expect you'd not want to "install" .pyx files in your CMakeLists, only the .so's need to be "installed". Ahh, you are probably trying to put them in-place, so they get auto-discovered. Not fond of that (mixing compiled and .py files in source expecting a different mixture in the wheel), but it should be possible with a few wheel ignores (**.pyx).

I agree that the pyx files don't need to be added at all. I didn't do anything special to add them, but we could try and clean up whatever in scikit-build-core is doing that. FWIW one hacky solution I tried was to replace the pyx file with the pxd file in the known source files provided to the `install` function, but that didn't fix Cython compilation which again suggests to me that Cython isn't actually using this finder.

**Comment by LecrisUT** [comment]
> Furthermore, pxd files would change the 1-1 mapping from module to source file into a many-to-one mapping. I've done that by manually modifying the installed `_pkg1_editable.py` file and it works in the specific test case

Seems you've identified the issue and it might be a way to approach it. I'm not familiar with cython, so I could use a breakdown of how you've edited, maybe a tree view of the files and the dictionary passed to the editable file that makes it work.

From what I read, we should filter the modules that have `.pyx` and not parse them in the editable file?

**Comment by vyasr** [comment]
Sorry, my wording was confusing. When I said "and it works in this specific case", I was referring to the fact that in general it would not be well-defined for the `known_source_files` argument to the `ScikitBuildRedirectingFinder` to be a `dict[str, list[str]]` instead of a `dict[str, str]` because there needs to be a unique file to associate with the module. However, in hindsight I think this is a non-issue because the reason you end up with multiple files in the current situation is that we include pyx files, and we really shouldn't be including them at all. So in a world where the module loader approach worked for Cython, I think just removing pyx files from the source list while including pxd files when they are available would be sufficient. However, as I said above I'm not yet convinced that the module loader approach is even valid for Cython.

> From what I read, we should filter the modules that have .pyx and not parse them in the editable file?

Essentially Cython can be thought of as a compiled superset of Python. In analogy to C, `.pyx` files are akin to `.c` source files, and `.pxd` files are akin to `.h` header files. Since every source file is compiled into a shared object, a given pyx file module's lookup in the redirecting finder should always be to an `.so` file in the `known_wheel_files` list. You should never include the `.pyx` files in the `known_source_files` list because `.pyx` files aren't directly importable; `known_source_files` should only contain `.py` files.

Cython supports `cimport` statements, which are basically import statements for other Cython modules. In order for one Cython module to use another one in certain contexts (basically any context that isn't pure Python) it has to `cimport` rather than `import` that module. However, Cython modules are not `cimport`able by default. To enable that, you have to put declarations into a corresponding `pxd` file (unlike C, this isn't an optional separation; the Cython compiler will not allow you to directly import from pyx at all, unlike a C compiler where this is possible but bad practice).

To debug, I've created a local environment with scikit-build-core, then done an editable install of pkg1 into it (which works fine). I then do an editable install of pkg2, which fails as expected. The following approaches all work to remedy this issue:
- Set the PYTHONPATH directly to the pkg1/src directory, i.e. `PYTHONPATH=/path/to/pkg1/src pip install -e . --no-build-isolation --config-settings=build_dir=build/{wheel_tag} -vv` in the pkg2 directory
- Add a `test.pth` file into my site-packages dir containing `/path/to/pkg1/src`
- Add a `test.pth` file into my site-packages dir containing `import mypath`, and then adding another file mypath.py to site-packages that just does `import sys;sys.path.append("/path/to/pkg1/src")`

However, if I add a `test.pth` file into my site-packages dir containing `import mypath` and then have `mypath.py` containing:
```
import importlib.abc
import sys


class Finder(importlib.abc.MetaPathFinder):
    def find_spec(self, fullname, path=None, target=None):
        print(f"find_spec: {fullname=}, {path=}, {target=}", file=sys.stderr)
        return None


sys.meta_path.insert(0, Finder())
```

None of the prints of "find_spec..." mention pkg1 during Cython compilation. All of the mentions that I do see are of Python builds or Cython internals, suggesting that since Cython is itself implemented in Python we're seeing the bits that imports in order to do the transpiling, but the transpiling itself does not respect modifications of `sys.meta_path`.

Does that make sense?

**Comment by vyasr** [comment]
All right, after some discussion at the last dev meeting we came up with an alternative solution that involves just adding the extra paths directly to the pth file. My diagnosis in https://github.com/scikit-build/scikit-build-core/pull/516#issuecomment-1797116250 is correct, and the new solution leverages the fact that Cython does support pth files for paths even though it does not support meta path finders. I will follow up with Cython to try and get that support added in the future, but this solution seems workable in scikit-build-core for now.

@henryiii I've added a test, but I'm not sure how to get it to pass. How should I get installation of scikit-build-core in the virtual environment to install the current local package instead of pulling from PyPI? We could invoke `pip install /path/to/scikit-build-core`, but I'm not sure if that is the best option or if this is a problem that's already solved in other tests that I'm not seeing.

File: src/scikit_build_core/build/wheel.py [hunk 1]
```diff
@@ -30,6 +30,7 @@
 
 if TYPE_CHECKING:
     from collections.abc import Sequence
+    from typing import Iterable
```
Comment (henryiii) [inline src/scikit_build_core/build/wheel.py:33]:
```suggestion
    from collections.abc import Iterable, Sequence
```

The typing imports are deprecated.

File: src/scikit_build_core/build/wheel.py [hunk 2]
```diff
@@ -68,9 +70,15 @@ def _make_editable(
         f"_{name}_editable.py",
         editable_txt.encode(),
     )
+    # Support Cython by adding the source directory directly to the path.
+    # This is necessary because Cython does not support sys.meta_path for
+    # cimports (as of 3.0.5).
+    pth_import_paths = (
+        f"import _{name}_editable\n" + "\n".join(pkg for pkg in packages) + "\n"
+    )
```
Comment (henryiii) [inline src/scikit_build_core/build/wheel.py:78]:
```suggestion
    import_strings = ["import _{name}_editable", *packages, ""]
    pth_import_paths = "\n".join(import_strings)
```

File: src/scikit_build_core/build/wheel.py [hunk 3]
```diff
@@ -68,9 +69,14 @@ def _make_editable(
         f"_{name}_editable.py",
         editable_txt.encode(),
     )
+    # Support Cython by adding the source directory directly to the path.
+    # This is necessary because Cython does not support sys.meta_path for
+    # cimports (as of 3.0.5).
+    import_strings = ["import _{name}_editable", *packages, ""]
```
Comment (henryiii) [inline src/scikit_build_core/build/wheel.py:75]:
```suggestion
    import_strings = [f"import _{name}_editable", *packages, ""]
```

**Comment by vyasr** [comment]
Thanks for the fix! I guess the isolated fixture and the associated pep518 wheelhouse are designed exactly for the use case of pulling the current version of the code, makes sense. Let me know if you need anything else from me on this PR.

**Comment by henryiii** [comment]
Yes, they are. :)

**Pull request merged**